firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java

```diff
@@ -485,14 +486,17 @@ private HttpURLConnection openHttpURLConnection(URL url, String apiKey)
     httpURLConnection.addRequestProperty(CONTENT_ENCODING_HEADER_KEY, GZIP_CONTENT_ENCODING);
     httpURLConnection.addRequestProperty(CACHE_CONTROL_HEADER_KEY, CACHE_CONTROL_DIRECTIVE);
     httpURLConnection.addRequestProperty(X_ANDROID_PACKAGE_HEADER_KEY, context.getPackageName());
-    if ((heartbeatInfo.get() != null) && (userAgentPublisher.get() != null)) {
-      HeartBeat heartbeat =
-          heartbeatInfo.get().getHeartBeatCode(FIREBASE_INSTALLATIONS_ID_HEARTBEAT_TAG);
-      if (heartbeat != HeartBeat.NONE) {
+    HeartBeatController heartBeatController = heartBeatProvider.get();
+    if (heartBeatController != null) {
+      try {
         httpURLConnection.addRequestProperty(
-            USER_AGENT_HEADER, userAgentPublisher.get().getUserAgent());
+            GMP_APP_HEADER, firebaseApp.getOptions().getApplicationId());
         httpURLConnection.addRequestProperty(
-            HEART_BEAT_HEADER, Integer.toString(heartbeat.getCode()));
+            HEART_BEAT_HEADER, Tasks.await(heartBeatController.getHeartBeatsHeader()));
+      } catch (ExecutionException e) {
+        Log.w(TAG, "Failed to get heartbeats header");
+      } catch (InterruptedException e) {
+        Log.w(TAG, "Failed to get heartbeats header");
```

Comment:
You should restore the interrupted status here, context go/java-practices/interruptedexception